logins-sql/src/schema.rs

```diff
@@ -0,0 +1,202 @@
+        "CREATE TABLE IF NOT EXISTS {meta} (
+            key TEXT PRIMARY KEY,
+            value NOT NULL
+        )",
+        meta = META_TABLE_NAME,
+    );
+
+    static ref CREATE_OVERRIDE_HOSTNAME_INDEX_SQL: String = format!(
+        "CREATE INDEX IF NOT EXISTS {idx_override_hostname} ON {mirror} (is_overridden, hostname)",
+        idx_override_hostname = IDX_MIRROR_OVERRIDEN_HOSTNAME,
+        mirror = MIRROR_TABLE_NAME
+    );
+
+    static ref CREATE_DELETED_HOSTNAME_INDEX_SQL: String = format!(
+        "CREATE INDEX IF NOT EXISTS {idx_local_deleted_hostname} ON {local} (is_deleted, hostname)",
+        idx_local_deleted_hostname = IDX_LOCAL_DELETED_HOSTNAME,
+        local = LOCAL_TABLE_NAME
+    );
+
+    static ref UPDATE_LOCAL_TIMESTAMPS_TO_MILLIS_SQL: String = format!(
+        "UPDATE {local}
+         SET timeCreated = timeCreated / 1000,
+             timeLastUsed = timeLastUsed / 1000,
+             timePasswordChanged = timePasswordChanged / 1000",
+        local = LOCAL_TABLE_NAME
+    );
+
+    static ref UPDATE_MIRROR_TIMESTAMPS_TO_MILLIS_SQL: String = format!(
+        "UPDATE {mirror}
+         SET timeCreated = timeCreated / 1000,
+             timeLastUsed = timeLastUsed / 1000,
+             timePasswordChanged = timePasswordChanged / 1000",
+        mirror = MIRROR_TABLE_NAME
+    );
+
+    static ref SET_VERSION_SQL: String = format!(
+        "PRAGMA user_version = {version}",
+        version = VERSION
+    );
+}
+
+pub(crate) static LAST_SYNC_META_KEY:    &'static str = "last_sync_time";
+pub(crate) static GLOBAL_STATE_META_KEY: &'static str = "global_state";
+
+pub fn init(db: &db::LoginDb) -> Result<()> {
+    let user_version = db.query_one::<i64>("PRAGMA user_version")?;
+    if user_version == 0 {
+        let table_list_exists = db.query_one::<i64>(
+            "SELECT count(*) FROM sqlite_master WHERE type = 'table' AND name = 'tableList'"
+        )? != 0;
+
+        if !table_list_exists {
+            return create(db);
+        }
+    }
+    if user_version != VERSION {
+        if user_version < VERSION {
+            upgrade(db, user_version)?;
+        } else {
+            warn!("Loaded future schema version {} (we only understand version {}). \
+                   Optimisitically ",
+                  user_version, VERSION)
+        }
+    }
+    Ok(())
+}
+
+// https://github.com/mozilla-mobile/firefox-ios/blob/master/Storage/SQL/LoginsSchema.swift#L100
+fn upgrade(db: &db::LoginDb, from: i64) -> Result<()> {
+    debug!("Upgrading schema from {} to {}", from, VERSION);
+    if from == VERSION {
+        return Ok(());
+    }
+    if from == 0 {
+        drop(db)?;
+        create(db)?;
+        return Ok(());
+    }
+    if from < 3 {
+        // These indices were added in v3 (apparently)
+        db.execute_all(&[
+            &*CREATE_OVERRIDE_HOSTNAME_INDEX_SQL,
+            &*CREATE_DELETED_HOSTNAME_INDEX_SQL,
+        ])?;
+    }
+    if from < 4 {
```

Comment:
4 is this implementation, right, so not present in Firefox for iOS?  Worth calling that out (if it's not been done elsewhere) -- I might find it as I read.